lib/Compiler.cpp

```diff
@@ -372,6 +372,52 @@ int ParseSamplerMap(const std::string &sampler_map,
   return 0;
 }
 
+clang::TargetInfo *PrepareTargetInfo(CompilerInstance &instance) {
+  // Create target info
+  auto TargetInfo = clang::TargetInfo::CreateTargetInfo(
+      instance.getDiagnostics(),
+      std::make_shared<clang::TargetOptions>(instance.getTargetOpts()));
+
+  // The SPIR target enables all possible options, disable the ones we don't
+  // want
+  auto &Opts = TargetInfo->getSupportedOpenCLOpts();
+
+  // Conditionally disable extensions based on support
+  if (!clspv::Option::FP16()) {
+    Opts["cl_khr_fp16"] = false;
+  }
+  if (!clspv::Option::FP64()) {
+    Opts["cl_khr_fp64"] = false;
+  }
+
+  // Disable CL3.0 feature macros for unsupported features
+  if (instance.getLangOpts().LangStd == clang::LangStandard::lang_opencl30) {
+
+    // TODO: See #705, find a better solution for this.
+    // TODO(kpet): This is a Clang bug (the pragma isn't enabled when the macro
+    // is defined)
+    Opts["cl_khr_3d_image_writes"] = false;
+
+    // The following features are never supported
+    Opts["__opencl_c_pipes"] = false;
+    Opts["__opencl_c_generic_address_space"] = false;
+    Opts["__opencl_c_atomic_order_seq_cst"] = false;
+    Opts["__opencl_c_device_enqueue"] = false;
+    Opts["__opencl_c_read_write_images"] = false;
```

Comment:
These should actually work. There is code elsewhere that assumes read/write images for >= 2.0

Reply:
Yes, you're right. I'd forgotten you'd added support.

Reply:
Now, updated. PTAL.